include/sycl/algorithm/generate.hpp

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2015 The Khronos Group Inc.
+
+   Permission is hereby granted, free of charge, to any person obtaining a
+   copy of this software and/or associated documentation files (the
+   "Materials"), to deal in the Materials without restriction, including
+   without limitation the rights to use, copy, modify, merge, publish,
+   distribute, sublicense, and/or sell copies of the Materials, and to
+   permit persons to whom the Materials are furnished to do so, subject to
+   the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Materials.
+
+   MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+   KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+   SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+    https://www.khronos.org/registry/
+
+  THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+  CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+  MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+*/
+
+#ifndef __SYCL_IMPL_ALGORITHM_GENERATE__
+#define __SYCL_IMPL_ALGORITHM_GENERATE__
+
+#include <algorithm>
+#include <iostream>
+#include <type_traits>
+
+// SYCL helpers header
+#include <sycl/helpers/sycl_buffers.hpp>
+
+namespace sycl {
+namespace impl {
+
+/* generate.
+ * Implementation of the command group that submits a generate kernel.
+ * The kernel is implemented as a lambda.
+ */
+template <class ExecutionPolicy, class ForwardIt, class Generator>
+void generate(ExecutionPolicy &sep, ForwardIt first, ForwardIt last,
+              Generator g) {
+  cl::sycl::queue q{sep.get_queue()};
+  const auto device = q.get_device();
+
+  auto bufI = helpers::make_buffer(first, last);
+  const auto vectorSize = bufI.get_count();
+
+  const auto localRange =
+      device.get_info<cl::sycl::info::device::max_work_group_size>();
+  const auto globalRange = sep.calculateGlobalSize(vectorSize, localRange);
+
+  const auto f = [vectorSize, localRange, globalRange, &bufI,
+            g](cl::sycl::handler &h) mutable {
+    cl::sycl::nd_range<1> r{
+        cl::sycl::range<1>{std::max(globalRange, localRange)},
+        cl::sycl::range<1>{localRange}};
+    const auto aI = bufI.template get_access<cl::sycl::access::mode::read_write>(h);
+    h.parallel_for<typename ExecutionPolicy::kernelName>(
+        r, [aI, g, vectorSize](cl::sycl::nd_item<1> id) {
+          if (id.get_global(0) < vectorSize) {
+            aI[id.get_global(0)] = g();
+          }
+        });
+  };
```

Comment:
This will only work on vectors of size up to the maximum number of work items of the device. I think all other algorithms similar to this one are done in the same way. Can we create a separate issue to implement the algorithms so they have multiple passes when there are more items than the number of work items?

Reply:
What do you mean? There is no maximum number of global work-items for a device is there?

Reply:
There is, and, in some devices, it may be smaller than you think.
From https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clGetDeviceInfo.html :

> 
Parameter | Description
-- | --
CL_DEVICE_MAX_WORK_ITEM_SIZES | Return type: size_t[]. Maximum number of work-items that can be specified in each dimension of the work-group to clEnqueueNDRangeKernel.Returns n size_t entries, where n is the value returned by the query for CL_DEVICE_MAX_WORK_ITEM_DIMENSIONS. The minimum value is (1, 1, 1).

Reply:
But isn't that for a single work-group, so the local size.
*Maximum number of work-items that can be specified in each dimension of the* **work-group**.

Reply:
You are right, wrong info parameter. Seems it is clearer in CUDA (http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities, Maximum x-dimension of a grid of thread blocks, and y and z too).
There is a device param info to query for the maximum global work size *only for built in kernels* (which got me confused):

cl_kernel_work_group_info | Return Type | Info. returned in param_value
-- | -- | --
CL_KERNEL_GLOBAL_WORK_SIZE | size_t[3] | This provides a mechanism for the application to query the maximum global size that can be used to execute a kernel (i.e. global_work_sizeargument to clEnqueueNDRangeKernel) on a custom device given by device or a built-in kernel on an OpenCL device given by device.If device is not a custom device or kernel is not a built-in kernel, clGetKernelArgInfo returns the error CL_INVALID_VALUE.

However, **there is a limit on work item sizes**: the maximum value of size_t *on the device*. From errors in https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clEnqueueNDRangeKernel.html:

> CL_INVALID_GLOBAL_WORK_SIZE if global_work_size is NULL, or if any of the values specified in global_work_size[0], ...global_work_size [work_dim - 1] are 0 or exceed the range given by the sizeof(size_t) for the device on which the kernel execution will be enqueued.

Where:

Type in OpenCL C Language | Description | API type for application
-- | -- | --
size_t | The unsigned integer type of the result of the sizeof operator. This is a 32-bit unsigned integer if CL_DEVICE_ADDRESS_BITS defined in clGetDeviceInfo is 32-bits and is a 64-bit unsigned integer if CL_DEVICE_ADDRESS_BITS is 64-bits. | n/a

It is particularly problematic in SYCL when your host is 64 bit and your device is 32, since the maximum array size you can use on the host does not match the device, and you may not get a nice error if that happens!

Reply:
Oooh, that could be very unexpected. I'll create the issue.

Reply:
Issue #51